src/Psalm/Internal/Provider/ReturnTypeProvider/InArrayReturnTypeProvider.php

```diff
@@ -0,0 +1,81 @@
+<?php
+namespace Psalm\Internal\Provider\ReturnTypeProvider;
+
+use PhpParser\Node\Expr\ConstFetch;
+use Psalm\Internal\Type\Comparator\UnionTypeComparator;
+use Psalm\Plugin\EventHandler\Event\FunctionReturnTypeProviderEvent;
+use Psalm\Plugin\EventHandler\FunctionReturnTypeProviderInterface;
+use Psalm\Type;
+use Psalm\Type\Atomic\TArray;
+use Psalm\Type\Atomic\TKeyedArray;
+use Psalm\Type\Atomic\TList;
+
+use function strtolower;
+
+class InArrayReturnTypeProvider implements FunctionReturnTypeProviderInterface
+{
+    /**
+     * @return array<lowercase-string>
+     */
+    public static function getFunctionIds(): array
+    {
+        return ['in_array'];
+    }
+
+    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): Type\Union
+    {
+        $call_args = $event->getCallArgs();
+
+        if (!isset($call_args[2])) {
+            return Type::getBool();
+        }
+
+        $third_arg = $call_args[2]->value;
+
+        if (!$third_arg instanceof ConstFetch
+            || strtolower($third_arg->name->parts[0]) !== 'true'
```

Comment:
This would work with literal `true` only, but is there any reason for that limitation? Can't we fetch the expression type from the node type provider and use that instead?

Reply:
> This would work with literal true only, but is there any reason for that limitation? Can't we fetch the expression type from the node type provider and use that instead?

I'm looking at the Psalm internals for the first time, and I forgot to come back to this bit after I discovered the node type provider. I'll change that right away!

> Not sure about all those changes to expected messages. Shouldn't we mark the return type with from_docblock flag when any of the source types are coming from docblock, like this:

Didn't know about that either, I'll try and see what I can do :+1:

Reply:
I think it's ok now, let me know if anything else can be improved